test/compiler_test.exs

```diff
@@ -0,0 +1,429 @@
+      end
+    end
+
+    test "computes forward pass with default options" do
+      for activation <- @activation_layers do
+        model = Axon.input({nil, 32}) |> Axon.activation(activation)
+        input = Nx.random_uniform({1, 32})
+
+        assert {_init_fn, predict_fn} = Axon.compile(model)
+        assert predict_fn.(%{}, input) == apply(Axon.Activations, activation, [input])
+      end
+    end
+
+    # test "computes forward pass with custom options" do
+    # end
+
+    test "computes forward pass with output policy" do
+      for activation <- @activation_layers do
+        model = Axon.input({nil, 32}) |> Axon.activation(activation)
+        policy = AMP.create_policy(output: {:bf, 16})
+        mp_model = AMP.apply_policy(model, policy)
+
+        assert {init_fn, predict_fn} = Axon.compile(mp_model)
+        assert Nx.type(predict_fn.(init_fn.(), Nx.random_uniform({1, 32}))) == {:bf, 16}
+      end
+    end
+  end
+
+  describe "dense" do
+    test "initializes in default case" do
+      model = Axon.input({nil, 32}) |> Axon.dense(1, name: "dense")
+
+      assert {init_fn, _predict_fn} = Axon.compile(model)
+      assert %{"dense_kernel" => kernel, "dense_bias" => bias} = init_fn.()
+      assert Nx.shape(kernel) == {32, 1}
+      assert Nx.type(kernel) == {:f, 32}
+      assert Nx.shape(bias) == {1}
+      assert Nx.type(bias) == {:f, 32}
+    end
+
+    test "initializes with custom initializers" do
+      model1 = Axon.input({nil, 32}) |> Axon.dense(1, name: "dense", kernel_initializer: :zeros)
+
+      assert {init_fn, _predict_fn} = Axon.compile(model1)
+      assert %{"dense_kernel" => kernel, "dense_bias" => bias} = init_fn.()
+      assert kernel == Axon.Initializers.zeros(shape: {32, 1})
+      assert Nx.shape(bias) == {1}
+      assert Nx.type(bias) == {:f, 32}
+
+      model2 = Axon.input({nil, 32}) |> Axon.dense(1, name: "dense", bias_initializer: :zeros)
+
+      assert {init_fn, _predict_fn} = Axon.compile(model2)
+      assert %{"dense_kernel" => kernel, "dense_bias" => bias} = init_fn.()
+      assert Nx.shape(kernel) == {32, 1}
+      assert Nx.type(kernel) == {:f, 32}
+      assert bias == Axon.Initializers.zeros(shape: {1})
+    end
+
+    test "computes forward pass" do
+      model = Axon.input({nil, 1}) |> Axon.dense(1, name: "dense", kernel_initializer: :identity)
+      input = Nx.iota({1, 1}, type: {:f, 32})
+
+      assert {init_fn, predict_fn} = Axon.compile(model)
+      assert %{"dense_kernel" => kernel, "dense_bias" => bias} = params = init_fn.()
+      assert predict_fn.(params, input) == Axon.Layers.dense(input, kernel, bias)
+    end
+
+    test "returns zero gradient for frozen parameters" do
+      model =
+        Axon.input({nil, 32})
+        |> Axon.dense(1, name: "dense")
+        |> Axon.freeze()
+
+      assert {init_fn, predict_fn} = Axon.compile(model)
+
+      backward = fn params, input ->
+        Nx.Defn.grad(params, &Nx.mean(predict_fn.(&1, input)))
+      end
```

Comment:
This could possibly be included as a function of the model similar to `Axon.predict` and `Axon.init`. Something like:

```elixir
model = ...
loss_fn = ...
gradients = Axon.backward(model, params, input, loss_fn)
```